ext/mysql2/client.c

```diff
@@ -121,52 +121,78 @@ struct nogvl_select_db_args {
 static VALUE rb_set_ssl_mode_option(VALUE self, VALUE setting) {
   unsigned long version = mysql_get_client_version();
 
-  if (version < 50630 || (version >= 50700 && version < 50703)) {
-    rb_warn( "Your mysql client library does not support setting ssl_mode; full support comes with 5.6.36+, 5.7.11+, 8.0+" );
+  /* Warn about versions that are known to be incomplete; these are pretty
+   * ancient, we want people to upgrade if they need SSL/TLS to work
+   *
+   * MySQL 5.x before 5.6.30 -- ssl_mode introduced but not fully working until 5.6.36)
+   * MySQL 5.7 before 5.7.3 -- ssl_mode introduced but not fully working until 5.7.11)
+   */
+  if ((version >= 50000 && version < 50630) || (version >= 50700 && version < 50703)) {
+    rb_warn("Your mysql client library version %lu does not support setting ssl_mode; full support comes with 5.6.36+, 5.7.11+, 8.0+", version );
     return Qnil;
   }
+
+  /* For these versions, map from the options we're exposing to Ruby to the constant available:
+   *   ssl_mode: :verify_identity to MYSQL_OPT_SSL_VERIFY_SERVER_CERT = 1
+   *   ssl_mode: :required to MYSQL_OPT_SSL_ENFORCE = 1
+   *   ssl_mode: :disabled to MYSQL_OPT_SSL_ENFORCE = 0
+   */
 #if defined(HAVE_CONST_MYSQL_OPT_SSL_VERIFY_SERVER_CERT) || defined(HAVE_CONST_MYSQL_OPT_SSL_ENFORCE)
   GET_CLIENT(self);
-  int val = NUM2INT( setting );
-  // Either MySQL 5.7.3 - 5.7.10, or Connector/C 6.1.3 - 6.1.x, or MariaDB 10.x and later
-  if ((version >= 50703 && version < 50711) || (version >= 60103 && version < 60200) || version >= 100000) {
+  int val = NUM2INT(setting);
+
+  /* Expected code path for MariaDB 10.x and MariaDB Connector/C 3.x
+   * Workaround code path for MySQL 5.7.3 - 5.7.10 and MySQL Connector/C 6.1.3 - 6.1.x
+   */
+  if (version >= 100000                         // MariaDB (all versions numbered 10.x)
+    || (version >= 30000 && version < 40000)    // MariaDB Connector/C (all versions numbered 3.x)
```

Comment:
I just double-checked the following definitions are available in the MariaDB connector C case.

* https://mariadb.com/kb/en/configuring-mariadb-connectorc-with-option-files/#ssl-verify-server-cert - MYSQL_OPT_SSL_VERIFY_SERVER_CERT - Introduced: MariaDB Connector/C 3.0.0 
* https://mariadb.com/kb/en/configuring-mariadb-connectorc-with-option-files/#ssl-enforce - MYSQL_OPT_SSL_ENFORCE - Introduced: MariaDB Connector/C 3.1.1

Reply:
I think the ifdefs will handle this as is, at the risk of someone compiling against 3.0.x but running against 3.1.x or vice versa.

Reply:
I think so. My comment above is not about the issue or not about asking to change the code. I just checked and noted here that the 2 definitions are available for the MariaDB Connector C version range: `version >= 30000 && version < 40000`.